Tech task 2

MAKABEO opened the pull request:
This PR introduces several key improvements to the project:

- Migrated the project from pure Java to Maven, enabling better dependency management.
- Integrated Lombok to reduce boilerplate code.
- Implemented Hibernate as the ORM and added JPA for database interaction.
- Configured PostgreSQL as the database.
- Added necessary libraries for unit testing, including JUnit and Mockito.
- Refactored the data access layer to follow the DAO pattern for better separation of concerns.
- Applied the Factory Method pattern to enhance object creation flexibility and reduce coupling.
- Implemented the Strategy pattern for generating reports dynamically.
- Introduced the Facade pattern to centralize component-related functionalities.
- Applied the Singleton pattern to ensure only one instance of certain services and facades.
- Introduced an enum to manage movie types efficiently.
- Implemented custom exceptions for better error handling and maintainability.
- Refactored the project to use EntityManager, decoupling the ORM usage.
- Enforced SOLID principles more extensively across the application.
- Added unit tests for services and StatementGenerator to improve test coverage.

MAKABEO committed:
Tech Task
- Updated Main.java
- Refactored Customer.java
- Refactored Movie.java
- Refactored Rental.java

MAKABEO committed:
Tech Task 2
- Created Custom exception

MAKABEO committed:
Tech Task 2
- Updated models with ID to prepare for database

MAKABEO committed:
Tech Task 2
- Implemented DAO

MAKABEO committed:
Tech Task 2
- Implemented Facade

MAKABEO committed:
Tech Task 2
- Created CustomerService and it interface
- Created RentalService and it interface
- Created MovieService and it interface

MAKABEO committed:
Tech Task 2
- Implemented Factory Method for facade and services

MAKABEO committed:
Tech Task 2
- Created RentalStatementCalculator to implement strategy

MAKABEO committed:
Tech Task 2
- Updated Main to add exception handling and use factory

MAKABEO committed:
Tech Task 2
- Changed project to work with maven

MAKABEO committed:
Tech Task 2
- Updated Customer to add lombok and use ORM
- Updated Movie to add lombok and use ORM
- Updated Rental to add lombok and use ORM
- Created HibernateUtil

MAKABEO committed:
Tech Task 2
- Updated .gitignore to permit .xml files
- Deleted HibernateUtil
- Added pom.xml to maven
- Added persistence.xml

MAKABEO committed:
Tech Task 2
- Created JPAUtil to create entityManager
- Created RentalMovie to handle manytomany relationship

MAKABEO committed:
Tech Task 2
- Updated CustomerDaoImpl to use EntityManager
- Updated CustomerFacade to add getCustomerByName
- Updated CustomerNotFoundException
- Updated CustomerServiceImpl to add getCustomerByName
- Updated ICustomerDao
- Updated ICustomerService

MAKABEO committed:
Tech Task 2
- Updated Movie to handle many to many relationship
- Updated MovieDaoImpl to use EntityManager

MAKABEO committed:
Tech Task 2
- Updated Rental to handle many to many relationship
- Updated RentalDaoImpl to use EntityManager
- Updated RentalService
- Updated RentalStatementCalculator and StatementGenerator.java to work with relationship

MAKABEO committed:
Tech Task 2
- Updated Main to handle the relationship

MAKABEO committed:
Tech Task 2
- Updated docker-compose to add pgadmin service
- Updated persistence.xml
- Updated pom.xml to add lombok

MAKABEO committed:
Tech Task 2
- Updated Main to delete unused import
- Updated pom.xml
- Added test for Services Implementations and StatementGeneratorTest

rcardoso-jalasoft commented on src/main/java/com/jaggaer/movies/model/enums/MovieType.java:
```diff
@@ -0,0 +1,45 @@
+    REGULAR {
+        @Override
+        public double calculateRentalAmount(int daysRented) {
+            double amount = 2;
+            if (daysRented > 2) {
+                amount += (daysRented - 2) * 1.5;
```
we should avoid hardcoding values

rcardoso-jalasoft commented on src/main/java/com/jaggaer/movies/dao/impl/CustomerDaoImpl.java:
```diff
@@ -0,0 +1,66 @@
+
+    @Override
+    public void save(Customer customer) {
+        entityManager.getTransaction().begin();
+        entityManager.persist(customer);
+        entityManager.getTransaction().commit();
```
what happens If an exception occurs between begin() and commit()? Use a try-catch block to properly handle rollbacks if needed

rcardoso-jalasoft commented on src/main/java/com/jaggaer/movies/dao/impl/MovieDaoImpl.java:
```diff
@@ -0,0 +1,52 @@
+
+    @Override
+    public void update(Movie movie) {
+        entityManager.getTransaction().begin();
+        entityManager.merge(movie);
+        entityManager.getTransaction().commit();
```
same use a try-catch block to properly handle rollbacks if needed

rcardoso-jalasoft commented on src/main/java/com/jaggaer/movies/dao/impl/RentalDaoImpl.java:
```diff
@@ -0,0 +1,52 @@
+    }
+
+    @Override
+    public void save(Rental rental) {
+        entityManager.getTransaction().begin();
+        entityManager.persist(rental);
+        entityManager.getTransaction().commit();
```
Same for all the class use a try-catch block to properly handle rollbacks

MAKABEO committed:
Tech Task 2
- Updated MovieType to avoid hardcoded values
- Created exceptions to handle persistence errors
- Updated dao implementations to handle persistence errors
- Updated facades to handle persistence errors
